jolt-core/src/jolt/vm/instruction_lookups.rs

```diff
@@ -988,85 +988,98 @@ where
-            .into_par_iter()
-            .map(|i| {
-                let access_sequence: &Vec<usize> = &subtable_lookup_indices[i];
-                DensePolynomial::from_usize(access_sequence)
-            })
-            .collect();
-
         let mut instruction_flag_bitvectors: Vec<Vec<u64>> =
             vec![vec![0u64; m]; Self::NUM_INSTRUCTIONS];
         for (j, op) in ops.iter().enumerate() {
             if let Some(op) = op {
                 instruction_flag_bitvectors[InstructionSet::enum_index(op)][j] = 1;
             }
         }
-
         let instruction_flag_polys: Vec<DensePolynomial<F>> = instruction_flag_bitvectors
             .par_iter()
             .map(|flag_bitvector| DensePolynomial::from_u64(flag_bitvector))
             .collect();
 
-        let mut lookup_outputs = Self::compute_lookup_outputs(&ops);
+        let log_M = M.log_2();
+        let (chunked_indices, mut lookup_outputs): (Vec<Vec<usize>>, Vec<F>) = ops
+            .par_iter()
+            .map(|op| {
+                if let Some(op) = op {
+                    (
+                        op.to_indices(C, log_M),
+                        F::from_u64(op.lookup_entry()).unwrap(),
+                    )
+                } else {
+                    (vec![0; C], F::zero())
+                }
+            })
+            .unzip();
         lookup_outputs.resize(m, F::zero());
         let lookup_outputs = DensePolynomial::new(lookup_outputs);
 
+        let (subtable_lookup_indices, dim): (Vec<Vec<usize>>, Vec<DensePolynomial<F>>) = (0..C)
+            .into_par_iter()
+            .map(|i| {
+                let mut access_sequence: Vec<usize> =
+                    chunked_indices.par_iter().map(|chunks| chunks[i]).collect();
+                access_sequence.resize(m, 0);
+                //TODO: remove clone
+                (
+                    access_sequence.clone(),
+                    DensePolynomial::from_usize(&access_sequence),
+                )
+            })
+            .collect();
+
+        let (read_cts, final_cts, E_polys): (
+            Vec<DensePolynomial<F>>,
+            Vec<DensePolynomial<F>>,
+            Vec<DensePolynomial<F>>,
+        ) = (0..preprocessing.num_memories)
+            .into_par_iter()
+            .fold(
+                || (Vec::new(), Vec::new(), Vec::new()),
+                |(mut read_acc, mut final_acc, mut E_acc), memory_index| {
+                    let dim_index = preprocessing.memory_to_dimension_index[memory_index];
+                    let subtable_index = preprocessing.memory_to_subtable_index[memory_index];
+                    let access_sequence: &Vec<usize> = &subtable_lookup_indices[dim_index];
+
+                    let mut final_cts_i = vec![0usize; M];
+                    let mut read_cts_i = vec![0usize; m];
+                    let mut subtable_lookups = vec![F::zero(); m];
+
+                    for (j, op) in ops.iter().enumerate() {
```

Comment:
It may be faster if instead of the current pattern:
```
for memory in memories {
    for op in ops {
         if(op.memories.contains(memory)) { // probably linear in the size of memories (4 -> 12 with current params)
                  // do stuff
         }
    }
}
```

Doing (bad psuedo code to show idea):
```
// Precompute list of instruction indices for each memory
let memory_A_op_indices = [];
let memory_B_op_indices = [];
let memory_ops = [memory_A_op_indices, memory_B_op_indices,...];
...
for op in ops {
     for memory in ops.memories_used {
          memory_ops[memory].push(op);
     }
} 

// Now compute the memory counters directly in parallel
for memory in memories {
                    let mut final_cts_i = vec![0usize; M];
                    let mut read_cts_i = vec![0usize; m];
                    let mut subtable_lookups = vec![F::zero(); m];

                  for op in memory_ops[memory] {
                       // update relevant counters
                  }

                 // TODO: construct polynomials
}
```

I'm thinking this will have better cache performance (less stuff is looked up in the hot loop) and get rid of the `memories_used.contains`.